badger/badger.go

```diff
@@ -229,14 +206,42 @@ type rangeIterator struct {
 	end      []byte
 	reverse  bool
 	keysOnly bool
-	closer   func() error
+	// reset is a mutatuble property that indicates whether the iterator should be
+	// returned to the beginning on the next [Next] call.
+	reset  bool
+	closer func() error
+}
+
+func (it *rangeIterator) Reset() {
+	it.reset = true
+}
+
+// restart returns the iterator back to it's initial location at time of construction,
+// allowing re-iteration of the underlying data.
+func (it *rangeIterator) restart() (bool, error) {
+	it.reset = false
+
+	if it.reverse {
+		hasValue, err := it.Seek(it.end)
+		if !hasValue || err != nil {
+			return false, err
+		}
+
+		if it.valid() && equal(it.i.Item().Key(), it.end) {
```

Comment:
suggesting: The call to `it.valid()` is redundant as `it.Seek()` already calls it and that is what `hasValue` represents.

Reply:
:grin: I have a memory of removing that, maybe I reverted it along with a broken change, re-removing is problem free, thanks Fred.

- [ ] Remove extra `valid` check